updatem/orchestration/update_orchestrator_test.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2023 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+
+package orchestration
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/eclipse-kanto/update-manager/api"
+	"github.com/eclipse-kanto/update-manager/api/types"
+	"github.com/eclipse-kanto/update-manager/config"
+	"github.com/eclipse-kanto/update-manager/test/mocks"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNewUpdateOrchestrator(t *testing.T) {
+	expectedOrchestrator := &updateOrchestrator{
+		cfg: &config.Config{
+			RebootEnabled: true,
+		},
+		phaseTimeout: 10 * time.Minute,
+	}
+	newUpdOrch := NewUpdateOrchestrator(&config.Config{RebootEnabled: true})
+
+	assert.Equal(t, expectedOrchestrator, newUpdOrch)
+}
+
+func TestUpdOrchApply(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	doneChan := make(chan bool, 1)
+	applyChan := make(chan bool, 1)
+
+	t.Run("test_valid_scenario", func(t *testing.T) {
+		eventCallback := mocks.NewMockUpdateManagerCallback(mockCtrl)
+		ctx := context.Background()
+
+		updOrchestrator := &updateOrchestrator{
+			cfg: createTestConfig(false, false),
+		}
+		desiredState := &types.DesiredState{
+			Domains: []*types.Domain{
+				{
+					ID: "domain1",
+				},
+			},
+		}
+		domainAgent1 := mocks.NewMockUpdateManager(mockCtrl)
+
+		domainAgent1.EXPECT().Apply(ctx, "testActivityId", desiredState).DoAndReturn(func(ctx context.Context, activityId string, desiredState *types.DesiredState) {
+			applyChan <- true
+		})
+		domainAgent1.EXPECT().Name().Times(2)
+
+		domainAgents := map[string]api.UpdateManager{
+			"domain1": domainAgent1,
+		}
+
+		eventCallback.EXPECT().HandleDesiredStateFeedbackEvent("device", "testActivityId", "", gomock.Any(), "", []*types.Action{}).Times(4)
+
+		go applyDesiredState(ctx, updOrchestrator, doneChan, domainAgents, "testActivityId", desiredState, eventCallback)
+
+		<-applyChan
+		updOrchestrator.HandleDesiredStateFeedbackEvent("domain1", "testActivityId", "", types.StatusIdentified, "", []*types.Action{})
+		updOrchestrator.HandleDesiredStateFeedbackEvent("domain1", "testActivityId", "", types.StatusCompleted, "", []*types.Action{})
+		updOrchestrator.HandleDesiredStateFeedbackEvent("domain1", "testActivityId", "", types.BaselineStatusCleanupSuccess, "", []*types.Action{})
+
+		<-doneChan
+
+	})
+	t.Run("test_empty_domainAgents_err_not_nil", func(t *testing.T) {
```

Comment:
Incase of error the `orchestrator.operation` is set as : ` &{ INCOMPLETE  map[] map[] <nil> map[] <nil> <nil>  <nil> <nil>  false <nil>}`, hence this test fails with:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
....
github.com/eclipse-kanto/update-manager/updatem/orchestration.(*updateOrchestrator).notifyFeedback(0x40001179b0, {0x25fa9d, 0xa}, {0x40001304c0, 0x40})
	/Users/gok2sf/go/src/github.com/eclipse-kanto/update-manager/updatem/orchestration/update_orchestrator_feedback.go:375 +0x100
github.com/eclipse-kanto/update-manager/updatem/orchestration.(*updateOrchestrator).Apply.func1()
	/Users/gok2sf/go/src/github.com/eclipse-kanto/update-manager/updatem/orchestration/update_orchestrator.go:62 +0x70
github.com/eclipse-kanto/update-manager/updatem/orchestration.(*updateOrchestrator).Apply(0x40001179b0, {0x2c9cd0, 0x4000124000}, 0x20db20?, {0x261393?, 0x20db20?}, 0x2c7ec8?, {0x2c83c8?, 0x4000113440?})
	/Users/gok2sf/go/src/github.com/eclipse-kanto/update-manager/updatem/orchestration/update_orchestrator.go:69 +0xdc
```

Is this expected and there is something wrong with the test or we found a bug? 
cc @stoyan-zoubev

Reply:
I haven't experienced this error here